File: app/src/main/java/com/karla/learningverbs/utils/firebase/CustomEventListener.java
package com.karla.learningverbs.utils.firebase;

import androidx.annotation.NonNull;

import com.karla.learningverbs.R;
import com.karla.learningverbs.utils.LearningApplication;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.ValueEventListener;

public abstract class CustomEventListener<T> implements ValueEventListener {
    private Class<T> customClass;

    public CustomEventListener(Class<T> GenericTypeIndicator) {
        showLoaging();
        this.customClass = GenericTypeIndicator;
    }

    @Override
    public void onDataChange(@NonNull DataSnapshot snapshot) {
        hideLoading();
        T currentCustomClass = null;
        for (DataSnapshot postSnapshot : snapshot.getChildren()) {
            currentCustomClass = postSnapshot.getValue(customClass);
        }
        if (currentCustomClass != null) {
            onSuccess(currentCustomClass);
        } else {
            onFailed(new Throwable(LearningApplication.getInstance().getString(R.string.msg_throwable)));
        }

    }

    @Override
    public void onCancelled(@NonNull DatabaseError error) {
        hideLoading();
        onFailed(new Throwable(error.getMessage()));
    }


    public abstract void onSuccess(T response);

    public abstract void onFailed(Throwable throwable);

    public abstract void showLoaging();

    public abstract void hideLoading();

}

File: app/src/main/java/com/karla/learningverbs/kotlin/utils/firebase/CustomEventListener.kt
package com.karla.learningverbs.kotlin.utils.firebase

import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.ValueEventListener
import com.karla.learningverbs.R
import com.karla.learningverbs.utils.LearningApplication2

abstract class CustomEventListener<T>(GenericTypeIndicator: Class<T>) :
    ValueEventListener {
    private val customClass: Class<T>

    init {
        showLoaging()
        customClass = GenericTypeIndicator
    }

    override fun onDataChange(snapshot: DataSnapshot) {
        hideLoading()
        var currentCustomClass: T? = null
        for (postSnapshot in snapshot.children) {
            currentCustomClass = postSnapshot.getValue(customClass)
        }
        if (currentCustomClass != null) {
            onSuccess(currentCustomClass)
        } else {
            onFailed(Throwable(LearningApplication2.getInstance().getString(R.string.msg_throwable)))
        }
    }

    override fun onCancelled(error: DatabaseError) {
        hideLoading()
        onFailed(Throwable(error.message))
    }

    abstract fun onSuccess(response: T)
    abstract fun onFailed(throwable: Throwable?)
    abstract fun showLoaging()
    abstract fun hideLoading()
}